app/models/loan_question.rb

```diff
@@ -181,6 +183,35 @@ def build_complete_requirements
     end
   end
 
+  def full_number
+    return @full_number if defined?(@full_number)
+    @full_number = if number.nil? || parent.nil?
+      nil
+    elsif parent.root?
+      number.to_s
+    else
+      "#{parent.full_number}.#{number}"
+    end
+  end
+
+  def full_number_and_label
+    [full_number, label].compact.join(". ")
+  end
+
+  protected
+
+  def set_numbers
+    # If parent_id is null (which can happen just after creation)
+    # there is no point in doing this.
+    return if parent_id.nil?
+    self.class.connection.execute("UPDATE loan_questions SET number = num FROM (
+      SELECT id, ROW_NUMBER() OVER (ORDER BY POSITION) AS num
+      FROM loan_questions
+      WHERE parent_id = #{parent_id} AND status = 'active'
+    ) AS t WHERE loan_questions.id = t.id")
```

Comment:
Wow, that's some fancy SQL. I guess efficiency is important here because it needs to run on drag and drop. Thanks for figuring this out!